harness/determined/pytorch/_pytorch_trial.py

```diff
@@ -116,15 +263,105 @@ def _set_random_seeds(cls: Type["PyTorchTrialController"], seed: int) -> None:
+            )
+
+        # Only report on the chief worker
+        if self._is_chief:
+            avg_metrics = metrics.get("avg_metrics", {})
+            batch_metrics = metrics.get("batch_metrics", [])
+
+            self._metric_writer.on_train_step_end(
+                self._state.batches_trained,
+                avg_metrics,
+                batch_metrics,
+            )
+            self._core_context.train.report_training_metrics(
+                steps_completed=self._state.batches_trained,
+                metrics=avg_metrics,
+                batch_metrics=batch_metrics,
+            )
+
+        self._context.reset_reducers()
+
+    def _is_best_validation(self, now: float, before: Optional[float]) -> bool:
+        if before is None:
+            return True
+
+        return (now < before) if self._smaller_is_better else (now > before)
+
+    def _on_epoch_start(self, epoch_idx: int):
+        for callback in self._callbacks.values():
+            with self._prof.record_timing(
+                f"callbacks.{callback.__class__.__name__}.on_training_epoch_start"
+            ):
+                sig = signature(callback.on_training_epoch_start)
+                if sig.parameters:
+                    callback.on_training_epoch_start(epoch_idx)
+                else:
+                    logging.warning(
+                        "on_training_epoch_start() without parameters is deprecated"
+                        " since 0.17.8. Please add epoch_idx parameter."
+                    )
+                    callback.on_training_epoch_start()  # type: ignore[call-arg]
+
+    def _on_epoch_end(self, epoch_idx: int):
+        for callback in self._callbacks.values():
+            with self._prof.record_timing(
+                f"callbacks.{callback.__class__.__name__}.on_training_epoch_end"
+            ):
+                callback.on_training_epoch_end(epoch_idx)
+
+    def _checkpoint(self, already_exiting: bool):
+        self._core_context.train.set_status("checkpointing")
+        self._state.last_ckpt = self._state.batches_trained
+
+        uuid = ""
+        try:
+            if self._is_chief:
+                metadata = {
+                    "determined_version": det.__version__,
+                    "steps_completed": self._state.batches_trained,
+                    "framework": f"torch-{torch.__version__}",
+                    "format": "pickle",
+                }
+                with self._context._core.checkpoint.store_path(metadata) as (
+                    path,
+                    storage_id,
+                ):
+                    self._save(path)
+                    uuid = storage_id
+            uuid = self._context.distributed.broadcast(uuid)
+            for callback in self._callbacks.values():
+                callback.on_checkpoint_upload_end(uuid=uuid)
+        except det.InvalidHP:
+            self._core_context.train.report_early_exit(core.EarlyExitReason.INVALID_HP)
```

Comment:
not sure where my earlier comment went, but this should be handled by `with det.core.init()` I think

Reply:
existing code checks for invalid HP, but the only meaningful thing it does with it is checkpoint before exiting. i see this check for invalidHP in core.__exit__, but we don't do anything if caught, we just exit. i can't think of a reason why invalidHP would occur anytime after initialization (during training), but maybe i'm missing something? if invalidHP can only occur on init, then we can get rid of these checks.

Reply:
discussed elsewhere